cmd/telemetry-bench.go

```diff
@@ -144,6 +144,32 @@ func (m *plugin) GetMetricMessage() (msgs []string) {
 	return buffers
 }
 
+//GetEventMessage generate mock collectd event messages
+func (m *plugin) GetEventMessage() (msg []string) {
+	bufferSize := len(m.mtype) * len(m.typeInstance) * len(m.pluginInstance)
+	buffers := make([]string, bufferSize)
+
+	msgMax := cap(m.mtype) * cap(m.pluginInstance) * cap(m.typeInstance)
+	for msgCount := 0; msgCount < msgMax; msgCount++ {
```

Comment:
Do you have a sample of a live message off of the AMQP bus that you based this on? I'd be interested to see it (or I'll capture one myself if you don't have one handy). 

AFAICT, like metrics, collectd notifications have plugin_instance and type_instance fields: https://github.com/collectd/collectd/blob/d96e10d68027716fb9d7f36a92db26f61145c493/src/daemon/plugin.h#L156 and if so I think you should populate them with iterated values like we do for metrics.

After a bit of digging, it looks like they would turn into part of the label data in a JSON formatted notification like this: "mangling of plugin instance and type instance into labels is copied from the Prometheus collectd exporter" https://github.com/collectd/collectd/blob/6b0e22a545bb3e4be1684a9fd6d2b48259ba9a7f/src/utils/format_json/format_json.c#L566 

I'm also going to admit that I'm a bit baffled why the use of cap() instead of len() here and in the original code, but I haven't gotten out the debugger yet to figure out what I'm missing. AFAICT these slices are filled with values like "type1", "typInst42", "pluginInst99" and len() should represent the number we are trying to generate. cap() may (or may not) be larger than len(). I feel silly even asking, but there it is. Can you explain?

Reply:
Thanks for reviewing!
Here is the raw message generated by the collectd interface plugin:
```
{
	"message": "Host saf-perftest-notify-f8825 (...) Current value of \"rx\" is 0.000000.",
	"host": "saf-perftest-notify-f8825",
	"plugin": "interface",
	"plugin_instance": "lo",
	"type": "if_octets",
	"severity": "ok",
	"level": "info",
	"@timestamp": "2019-11-25T18:36:35Z"
}
```
I don't have an event directly from AMQ untouched by SG, but based on [SG event code](https://github.com/redhat-service-assurance/smart-gateway/blob/68f717e8caf0ac356f754c31115f4dfcea449cff/internal/pkg/events/events.go#L203), SG merely utilizes some of the fields to specify indices for ElasticSearch and writes the raw messages as documents in that index. Here is a raw message posted to EasticSearch by SG.

```
{
	"labels": {
		"alertname": "collectd_interface_if_octets",
		"instance": "saf-perftest-notify-f8825",
		"interface": "lo",
		"severity": "OKAY",
		"service": "collectd"
	},
	"annotations": {
		"summary": "Host saf-perftest-notify-f8825, plugin interface (instance lo) type if_octets: All data sources are within range again. Current value of \"rx\" is nan.",
		"DataSource": "rx",
		"CurrentValue": "nan",
		"WarningMin": "nan",
		"WarningMax": "nan",
		"FailureMin": "nan",
		"FailureMax": "0"
	},
	"startsAt": "2019-11-25T18:35:28.432871415Z"
}
```
The collectd code appears to add the plugin name and plugin instance as its own key value pair based on the "mangling of plugin instance ..." section you reference above. In the sample event, this would be the `"interface":"lo"` key-value pair. In the PR, I placed the plugin name into the "service" field which collectd [manually sets](https://github.com/collectd/collectd/blob/6b0e22a545bb3e4be1684a9fd6d2b48259ba9a7f/src/utils/format_json/format_json.c#L587) to "collectd." For authenticity, TB should probably follow the same pattern and I will adjust that. Otherwise, "alertname","instance","severity", and "service" are explicitly set by the collectd json plugin.

For annotations, the only thing collectd specifies is the "summary" section, which I omitted only to keep the code looking cleaner.

Realisticly, if we want to have plugin_instance and type_instance fields, TB can add those without any problems. If the information would be useful for testing I can add them into the annotations section.

As for why the original code chose to use cap() instead of len(), I am not sure. Based on my testing and observation, it doesn't *seem* to make a difference because the generateHosts() function never allocates a larger buffer than what is needed for generating the number of types, plugin instances and type instances the tester specifies. I really only used it because the original function did

Reply:
Thanks for this excellent reply, Paul, I'm convinced! No need to add additional fields at this point, but in addition to the change of "service" I think you SHOULD actually include a summary of roughly an average length as this will affect the size of the event on the wire and in ES. Not a bit deal, though, you've addressed my major concerns.

Reply:
Will do! I can think of a few sarcastic messages to use as filler space 😀